File: Algoexpert/Diameter of a Binary Tree/cpp/diameter_of_a_binary_tree.cpp
#include <bits/stdc++.h>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {

public:
    int ans=0;

    int height(TreeNode* root) {
        if(!root) return 0;
        
        int lHeight = height(root->left);
        int rHeight = height(root->right);
        
        ans= max(ans, 1 + lHeight + rHeight);
        return 1+ max( lHeight , rHeight);

    }

    int diameterOfBinaryTree(TreeNode* root) {
        if(!root) return 0;
        height(root);
        return ans-1;
    }

};


int main() {

    TreeNode node1 = TreeNode(4);
    TreeNode node2 = TreeNode(2);
    TreeNode node3 = TreeNode(7);
    TreeNode node4 = TreeNode(1);
    TreeNode node5 = TreeNode(3);

    node1.left = &node2;
    node1.right = &node3;
    node2.left = &node4;
    node2.right = &node5;

    Solution solution;

    cout << solution.diameterOfBinaryTree(&node1) << endl;

    return 0;
}

File: LeetCode/Diameter of a Binary Tree/java/DiameterOfABinaryTree.java
public class DiameterOfABinaryTree {

    private static int diameter = 0;
    public static int diameterOfBinaryTree(TreeNode root) {
        diameterOfBinaryTreeHelper(root);
        return diameter;
    }

    public static int diameterOfBinaryTreeHelper(TreeNode root) {

        if(root==null) {
            return 0;
        }

        int leftHeight = diameterOfBinaryTreeHelper(root.left);
        int rightHeight = diameterOfBinaryTreeHelper(root.right);
        diameter = Math.max(leftHeight+rightHeight, diameter);

        return Math.max(leftHeight,rightHeight) + 1;

    }

    public static void main(String[] args) {
        TreeNode node1 = new TreeNode(4);
        TreeNode node2 = new TreeNode(2);
        TreeNode node3 = new TreeNode(7);
        TreeNode node4 = new TreeNode(1);
        TreeNode node5 = new TreeNode(3);

        node1.left = node2;
        node1.right = node3;
        node2.left = node4;
        node2.right = node5;

        System.out.println(diameterOfBinaryTree(node1));

    }

}
